docs/architecture/adr-038-state-listening.md

```diff
@@ -0,0 +1,584 @@
+# ADR 038: KVStore state listening
+
+## Changelog
+
+- 11/23/2020: Initial draft
+
+## Status
+
+Proposed
+
+## Abstract
+
+This ADR defines a set of changes to enable listening to state changes of individual KVStores and exposing these data to consumers.
+
+## Context
+
+Currently, KVStore data can be remotely accessed through [Queries](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/messages-and-queries.md#queries) which proceed through Tendermint and the ABCI.
+In addition to these request/response queries, it would be beneficial to have a means of listening to state changes as they occur in real time.
+
+## Decision
+
+We will modify the `MultiStore` interface and its concrete (`rootmulti` and `cachemulti`) implementations and introduce a new `listenkv.Store` to allow listening to state changes in underlying KVStores.
+We will also introduce the tooling for writing these state changes out to a file.
+
+### Listening interface
+
+In a new file- `store/types/listening.go`- we will create a `WriteListener` interface for streaming out state changes from a KVStore.
+
+```go
+// WriteListener interface for streaming data out from a listenkv.Store
+type WriteListener interface {
+  // if value is nil then it was deleted
+  OnWrite(key []byte, value []byte)
```

Comment:
would the block height be useful here?

Reply:
I think it would, but I don't think it is possible to emit that information from within the context of the KVStores without some significant refactoring.

Reply:
Can we not pass in an `sdk.Context` as part of this? That way we can grab those fields (i.e the block header).

Reply:
I don't think this is possible without some significant refactoring, since this method will be called from within a KVStore which do not (currently) hold a reference to `sdk.Context`.